File: src/cli/ls.rs
//! List workspaces command

use clap::Args;
use colored::Colorize;

use crate::core::{GitClient, WorkspaceManager, WorktreeManager};

#[derive(Args, Debug)]
pub struct LsArgs {
    /// Show detailed information including repo status
    #[arg(short, long)]
    long: bool,

    /// Output only workspace names (for shell completion)
    #[arg(short, long)]
    short: bool,
}

pub async fn execute(args: LsArgs, manager: WorkspaceManager) -> anyhow::Result<()> {
    let workspaces = manager.list_workspaces();

    if workspaces.is_empty() {
        if !args.short {
            println!("{}", "No workspaces found.".dimmed());
            println!();
            println!("Create a workspace with:");
            println!("  {}", "wtp create <workspace_name>".cyan());
            println!();
            println!("All workspaces are stored under workspace_root (default: ~/.wtp/workspaces)");
        }
        return Ok(());
    }

    if args.short {
        // Short format - just names, one per line (for shell completion)
        for ws in workspaces {
            println!("{}", ws.name);
        }
    } else if args.long {
        // Detailed listing with repo details
        let git = GitClient::new();

        for (i, ws) in workspaces.iter().enumerate() {
            if i > 0 {
                println!();
            }

            if !ws.exists || !ws.path.join(".wtp").exists() {
                println!(
                    "{}  {}",
                    ws.name.cyan().bold(),
                    "[missing]".red()
                );
                continue;
            }

            println!("{}", ws.name.cyan().bold());

            match WorktreeManager::load(&ws.path) {
                Ok(wt_manager) => {
                    let worktrees = wt_manager.list_worktrees();
                    if worktrees.is_empty() {
                        println!("  {}", "(no repos)".dimmed());
                    } else {
                        for wt in worktrees {
                            let wt_full_path = ws.path.join(&wt.worktree_path);
                            let repo_display = wt.repo.display();

                            if !wt_full_path.exists() {
                                println!(
                                    "  {:<30} {:<20} {}",
                                    repo_display,
                                    wt.branch.cyan(),
                                    "? missing".red()
                                );
                                continue;
                            }

                            let status_str = match git.get_status(&wt_full_path) {
                                Ok(s) => s.format_compact(),
                                Err(_) => "?".to_string(),
                            };

                            println!(
                                "  {:<30} {:<20} {}",
                                repo_display,
                                wt.branch.cyan(),
                                status_str
                            );
                        }
                    }
                }
                Err(_) => {
                    println!("  {}", "(error loading worktrees)".red());
                }
            }
        }
    } else {
        // Default listing: name + repo count
        for ws in workspaces {
            let name = ws.name.cyan().bold().to_string();

            if !ws.exists || !ws.path.join(".wtp").exists() {
                println!("{}  {}", name, "[missing]".red());
                continue;
            }

            let repo_info = match WorktreeManager::load(&ws.path) {
                Ok(wt_manager) => {
                    let count = wt_manager.list_worktrees().len();
                    match count {
                        0 => "(no repos)".to_string(),
                        1 => "(1 repo)".to_string(),
                        n => format!("({} repos)", n),
                    }
                }
                Err(_) => "(error)".to_string(),
            };

            println!("{}  {}", name, repo_info.dimmed());
        }
    }

    Ok(())
}

File: go/cmd/ls.go
package cmd

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/eddix/wtp/internal/core"
	"github.com/fatih/color"
	"github.com/spf13/cobra"
)

// NewLsCmd returns the "wtp ls" command.
func NewLsCmd() *cobra.Command {
	var long bool
	var short bool

	cmd := &cobra.Command{
		Use:     "ls",
		Short:   "List all workspaces",
		GroupID: GroupWorkspace,
		Args:    cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			loadedConfig, warning, err := core.LoadConfig()
			if err != nil {
				return err
			}
			if warning != "" {
				color.Yellow("%s", warning)
			}

			manager := core.NewWorkspaceManager(loadedConfig)
			workspaces := manager.ListWorkspaces()

			if len(workspaces) == 0 {
				if !short {
					dim := color.New(color.Faint)
					cyan := color.New(color.FgCyan)
					dim.Println("No workspaces found.")
					fmt.Println()
					fmt.Println("Create a workspace with:")
					fmt.Print("  ")
					cyan.Println("wtp create <workspace_name>")
					fmt.Println()
					fmt.Println("All workspaces are stored under workspace_root (default: ~/.wtp/workspaces)")
				}
				return nil
			}

			if short {
				return lsShort(workspaces)
			}
			if long {
				return lsLong(workspaces)
			}
			return lsDefault(workspaces)
		},
	}

	cmd.Flags().BoolVarP(&long, "long", "l", false, "Show detailed information including repo status")
	cmd.Flags().BoolVarP(&short, "short", "s", false, "Output only workspace names (for shell completion)")

	return cmd
}

func lsShort(workspaces []core.WorkspaceInfo) error {
	for _, ws := range workspaces {
		fmt.Println(ws.Name)
	}
	return nil
}

func lsDefault(workspaces []core.WorkspaceInfo) error {
	cyan := color.New(color.FgCyan, color.Bold)
	dim := color.New(color.Faint)
	red := color.New(color.FgRed)

	for _, ws := range workspaces {
		name := cyan.Sprintf("%s", ws.Name)

		if !ws.Exists || !dirHasWtp(ws.Path) {
			fmt.Printf("%s  ", name)
			red.Println("[missing]")
			continue
		}

		repoInfo := "(error)"
		mgr, err := core.LoadWorktreeManager(ws.Path, nil)
		if err == nil {
			count := len(mgr.ListWorktrees())
			switch count {
			case 0:
				repoInfo = "(no repos)"
			case 1:
				repoInfo = "(1 repo)"
			default:
				repoInfo = fmt.Sprintf("(%d repos)", count)
			}
		}

		fmt.Printf("%s  ", name)
		dim.Println(repoInfo)
	}
	return nil
}

func lsLong(workspaces []core.WorkspaceInfo) error {
	git := core.NewGitClient()
	cyan := color.New(color.FgCyan, color.Bold)
	red := color.New(color.FgRed)
	dim := color.New(color.Faint)

	for i, ws := range workspaces {
		if i > 0 {
			fmt.Println()
		}

		if !ws.Exists || !dirHasWtp(ws.Path) {
			cyan.Print(ws.Name)
			fmt.Print("  ")
			red.Println("[missing]")
			continue
		}

		cyan.Println(ws.Name)

		mgr, err := core.LoadWorktreeManager(ws.Path, nil)
		if err != nil {
			fmt.Print("  ")
			red.Println("(error loading worktrees)")
			continue
		}

		worktrees := mgr.ListWorktrees()
		if len(worktrees) == 0 {
			fmt.Print("  ")
			dim.Println("(no repos)")
			continue
		}

		for _, wt := range worktrees {
			wtFullPath := filepath.Join(ws.Path, wt.WorktreePath)
			repoDisplay := wt.Repo.Display()

			if !pathExists(wtFullPath) {
				fmt.Printf("  %-30s %-20s ", repoDisplay, color.CyanString("%s", wt.Branch))
				red.Println("? missing")
				continue
			}

			statusStr := "?"
			if status, err := git.GetStatus(wtFullPath); err == nil {
				statusStr = status.FormatCompact()
			}

			fmt.Printf("  %-30s %-20s %s\n", repoDisplay, color.CyanString("%s", wt.Branch), statusStr)
		}
	}
	return nil
}

// dirHasWtp checks if a directory has a .wtp subdirectory.
func dirHasWtp(path string) bool {
	return pathIsDir(filepath.Join(path, core.WtpDir))
}

// pathExists checks if a path exists.
func pathExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

// pathIsDir checks if a path exists and is a directory.
func pathIsDir(path string) bool {
	info, err := os.Stat(path)
	return err == nil && info.IsDir()
}

func init() {
	rootCmd.AddCommand(NewLsCmd())
}
